### PR TITLE
Indian GST accounting improvements

### PR DESCRIPTION
**Pull request by jva-odoo**
Task: https://www.odoo.com/web#id=42573&action=327&model=project.task&view_type=form&menu_id=4720

Pad : https://pad.odoo.com/p/r.8997df861854e285a816d7a7308fbf0c

### CODE REVIEW

File: addons/l10n_in/models/account.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    _inherit = "account.journal"
+
+    # Use for filter import and export type.
+    l10n_in_import_export = fields.Boolean("Import/Export", help="Tick this if this journal is use for Import/Export Under Indian GST.")
+
+
+class AccountMove(models.Model):
+    _inherit = "account.move"
+
+    l10n_in_gstin_partner_id = fields.Many2one(
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/l10n_in/models/account.py]: nope, this field value come either from invoice or payment or pos

File: addons/l10n_in/models/account_invoice.py [hunk 2]
```diff
@@ -28,16 +70,114 @@ def _get_report_base_filename(self):
...
+    def _onchange_partner_id(self):
+        if not self.env.context.get('from_purchase_order_change'):
+            if self.partner_id.state_id.country_id.code == 'IN':
+                self.l10n_in_place_of_supply = self.partner_id.state_id
+            else:
+                self.l10n_in_place_of_supply = self.env.ref('l10n_in.state_in_ot')
+        return super(AccountInvoice, self.with_context(l10n_in_gstin_partner_id=self.l10n_in_gstin_partner_id.id))._onchange_partner_id()
+
+
+class AccountInvoiceLine(models.Model):
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/l10n_in/models/account_invoice.py]: this is account.invoice.line and in account.py there is account.move.line

File: addons/l10n_in/models/account_invoice.py [hunk 3]
```diff
@@ -33,11 +75,141 @@ def _invoice_line_tax_values(self):
...
+        res = super(AccountInvoiceLine, self)._onchange_product_id()
+        self.l10n_in_is_eligible_for_itc = self.product_id.l10n_in_is_eligible_for_itc
+        return res
+
+
+class AccountInvoiceTax(models.Model):
+    _inherit = "account.invoice.tax"
+
+    l10n_in_product_id = fields.Many2one('product.product', string='Product')
+    l10n_in_uom_id = fields.Many2one('uom.uom', string='Unit of Measure')
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/l10n_in/models/account_invoice.py]: right, it was a mistake, added quantity as well.

File: addons/l10n_in/models/product_template.py [hunk 4]
```diff
@@ -8,3 +8,6 @@ class ProductTemplate(models.Model):
     _inherit = 'product.template'
 
     l10n_in_hsn_code = fields.Char(string="HSN/SAC Code", help="Harmonized System Nomenclature/Services Accounting Code")
+    l10n_in_hsn_description = fields.Char(string="HSN/SAC Description", help="HSN/SAC description is required if HSN/SAC code is not provided.")
+    l10n_in_is_eligible_for_itc = fields.Boolean(string="Is eligible for ITC",  help="Check this box if this product is eligible for ITC(Input Tax Credit) under GST")
+    l10n_in_is_asset = fields.Boolean(string="Is Asset", help="Check this box if this product is asset")
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/l10n_in/models/product_template.py]: they were related to GSTR2, hence removed for now

File: addons/l10n_in_pos/models/pos_order.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        l10n_in_place_of_supply = self.config_id.l10n_in_place_of_supply or self.env.ref('l10n_in.state_in_ot')
+        res['l10n_in_gstin_partner_id'] = self.config_id.l10n_in_gstin_partner_id.id
+        res['l10n_in_place_of_supply'] = l10n_in_place_of_supply and l10n_in_place_of_supply.id
+        return res
+
+
+class PosOrderLine(models.Model):
+    _inherit = "pos.order.line"
+
+    price_subtotal = fields.Float(compute='_compute_amount_line_all', digits=0, string='Subtotal w/o Tax', store=True)
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/l10n_in_pos/models/pos_order.py]: we do not need this in our PR now, do you still want me to make this change in POS module ?

File: addons/l10n_in/models/account_invoice.py [hunk 6]
```diff
@@ -14,6 +14,38 @@ def _compute_amount_total_words(self):
...
+        default=lambda self: self.env['res.company']._company_default_get('account.invoice').partner_id,
+        readonly=True, states={'draft': [('readonly', False)]}
+        )
+    l10n_in_partner_vat = fields.Char(related="partner_id.vat", readonly=True)
+    l10n_in_place_of_supply = fields.Many2one(
+        'res.country.state', string="Place Of Supply",readonly=True,
+        states={'draft': [('readonly', False)]}, domain=[("country_id.code", "=", "IN")])
+
+    @api.model
+    def _get_refund_common_fields(self):
```
Comment (antonylesuisse) [inline addons/l10n_in/models/account_invoice.py]: remove

File: addons/l10n_in/views/report_templates.xml [hunk 7]
```diff
@@ -0,0 +1,129 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
```
[the remark this answers is not in the document]
Comment (qdp-odoo) [reply addons/l10n_in/views/report_templates.xml]: we won't need this if we don't support the gstin_partner_id at first

File: addons/l10n_in/models/account_invoice.py [hunk 8]
```diff
@@ -14,6 +14,38 @@ def _compute_amount_total_words(self):
...
+    l10n_in_reseller_partner_id = fields.Many2one('res.partner', 'Reseller', domain=[('vat', '!=', False)], help="Only Registered Reseller", readonly=True, states={'draft': [('readonly', False)]})
+    l10n_in_gstin_partner_id = fields.Many2one(
+        'res.partner',
+        string="GSTIN",
+        required=True,
+        default=lambda self: self.env['res.company']._company_default_get('account.invoice').partner_id,
+        readonly=True, states={'draft': [('readonly', False)]}
+        )
+    l10n_in_partner_vat = fields.Char(related="partner_id.vat", readonly=True)
+    l10n_in_place_of_supply = fields.Many2one(
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/l10n_in/models/account_invoice.py]: As you say partner_shipping_id is coming from the sale module but I need this in l10n_in it only depends on the account module so I need re-introduced partner_shipping_id in module l10n_in and model account.invoice. so It's ok?
Also, need partner_shipping_id into account.move and pass the value from account.invoice to account.move.

File: addons/account/models/account_invoice.py [hunk 9]
```diff
@@ -1082,8 +1085,7 @@ def invoice_line_move_line_get(self):
             for tax in line.invoice_line_tax_ids:
                 tax_ids.append((4, tax.id, None))
                 for child in tax.children_tax_ids:
-                    if child.type_tax_use != 'none':
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/account/models/account_invoice.py]: Already dedicated commit for this.
https://github.com/odoo/odoo/pull/26262/commits/f691fc8077c6c319de3a421e0e8225e7e50da111

File: addons/point_of_sale/models/pos_order.py [hunk 10]
```diff
@@ -158,10 +158,15 @@ def _prepare_analytic_account(self, line):
...
-    def _create_account_move(self, dt, ref, journal_id, company_id):
-        date_tz_user = fields.Datetime.context_timestamp(self, fields.Datetime.from_string(dt))
+    def _create_account_move(self):
+        self.ensure_one()
+        date_tz_user = fields.Datetime.context_timestamp(self, fields.Datetime.from_string(self.session_id.start_at))
         date_tz_user = fields.Date.to_string(date_tz_user)
-        return self.env['account.move'].sudo().create({'ref': ref, 'journal_id': journal_id, 'date': date_tz_user})
+        return self.env['account.move'].sudo().create({
+            'ref': self.name,
+            'journal_id': self.sale_journal.id,
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/point_of_sale/models/pos_order.py]: Old:- account.move create from given peramiters like (dt, ref, journal_id, company_id)

New:- Value need to create account.move is available in self so we use self object values. so if have new fields in the related document then we can pass the value easily with inheritance.

[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/point_of_sale/models/pos_order.py]: ` journal_id = self.env['ir.config_parameter'].sudo().get_param(
     pos.closing.journal_id_%s' % current_company.id, default=order.sale_journal.id)`
this is use in odoo - v9 so we find in master that this parms is not set in odoo anywhere. so we   remove it.
it is ok or not.? give your valuable feedback.
you can see this change in https://github.com/odoo/odoo/pull/13989/commits/e99e0d6922742500ecb222b7667a85fe9220863f

File: addons/point_of_sale/models/pos_session.py [hunk 11]
```diff
@@ -19,15 +19,11 @@ class PosSession(models.Model):
 
     def _confirm_orders(self):
         for session in self:
-            company_id = session.config_id.journal_id.company_id.id
-            orders = session.order_ids.filtered(lambda order: order.state == 'paid')
-            journal_id = self.env['ir.config_parameter'].sudo().get_param(
-                'pos.closing.journal_id_%s' % company_id, default=session.config_id.journal_id.id)
-            if not journal_id:
+            journal = session.config_id.journal_id
```
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/point_of_sale/models/pos_session.py]: `pos.closing.journal_id_%s`
is use in odoo - v9 and we find in master that this parms is not set in odoo anywhere. so we remove it.
it is ok or not.? give your valuable feedback.

**Comment by qdp-odoo** [comment]
@jva-odoo can you give me the status of this task? are you still working on it?

File: addons/l10n_in/models/account_invoice.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (jva-odoo) [reply addons/l10n_in/models/account_invoice.py]: I discuss with @mga-odoo sir related to place_of_supply so now we decide to use the partner_id state as a place of supply.

so this field should be removed as you say.

File: addons/point_of_sale/models/pos_order.py [hunk 12]
```diff
@@ -253,6 +258,79 @@ def _flatten_tax_and_children(taxes, group_done=None):
...
+                    'partner_id': partner_id,
+                    'order_id': order.id
+                }
+                if currency_id != cur_company:
+                    data['currency_id'] = currency_id.id
+                    data['amount_currency'] = -abs(tax['amount']) if data.get('credit') else abs(tax['amount'])
+                res.append({'data_type': 'tax', 'values': data})
+        return res
+
+    def _create_account_move_line(self, session=None, move=None):
```
[the remark this answers is not in the document]
Comment (qdp-odoo) [reply addons/point_of_sale/models/pos_order.py]: I'm ok to do it in another task. Please create it in RD india and ping me on when it's done

**Comment by jva-odoo** [comment]
> @jva-odoo can you give me the status of this task? are you still working on it?

Yes, Work in progress. I solve most of the issues as per your comments.

**Comment by qdp-odoo** [comment]
closed with https://github.com/odoo/odoo/commit/451fd6cc90aaa14816376a0285334678df9ce484

**Pull request closed without merging**